client/verta/tests/custom_modules/test_custom_modules.py

```diff
@@ -0,0 +1,128 @@
+from verta._internal_utils.custom_modules import CustomModules
+
+from .. import strategies, utils
+from . import contexts
+
+
+class TestCollectPipInstalledModule:
+    @staticmethod
+    def assert_in_custom_modules(custom_modules, module_name):
+        module = CustomModules.get_module_path(module_name)
+
+        with utils.tempdir() as custom_modules_dir:
+            with zipfile.ZipFile(custom_modules, "r") as zipf:
+                zipf.extractall(custom_modules_dir)
+
+            # TODO: extract sys.path from _verta_config.py instead of walking
+            for parent_dir, dirnames, filenames in os.walk(custom_modules_dir):
+                if os.path.basename(module) in dirnames + filenames:
+                    retrieved_module = os.path.join(
+                        parent_dir,
+                        os.path.basename(module),
+                    )
+                    break
+            else:
+                raise ValueError("module not found in custom modules")
+
+            if os.path.isfile(module):
+                assert filecmp.cmp(module, retrieved_module)
+            else:
+                utils.assert_dirs_match(module, retrieved_module)
+
+    @pytest.mark.parametrize(
+        "name",
+        sorted(module[1] for module in pkgutil.iter_modules()),
+    )
+    def test_module(self, name):
+        """pip-installed module can be collected."""
+        if name == "tests" or name == "conftest" or name.startswith("test_"):
+            pytest.skip(
+                "pytest modifies both import mechanisms and module objects,"
+                " which we can't handle right now"
+            )
+        if six.PY2 and name == "pytest_forked":
+            pytest.skip(
+                "pytest_forked insists on having an empty __pycache__,"
+                " which custom modules ignores, which fails our match check"
+            )
+        if CustomModules.get_module_path(name) in ("built-in", "frozen"):
+            pytest.skip("built into Python; no module file to collect")
+        if six.PY2 and (name.startswith("tensorflow_") or name == "torch"):
+            pytest.skip("takes too long")
+
+        custom_modules = _DeployableEntity._custom_modules_as_artifact([name])
+        self.assert_in_custom_modules(custom_modules, name)
+
+    @pytest.mark.parametrize(
+        "names",
+        [
+            ["cloudpickle", "hypothesis"],
+            ["cloudpickle", "hypothesis", "pytest"],
+        ],
+    )
+    def test_multiple_modules(self, names):
+        """Multiple pip-installed modules can be collected at once."""
+        custom_modules = _DeployableEntity._custom_modules_as_artifact(names)
+        for name in names:
+            self.assert_in_custom_modules(custom_modules, name)
+
+    def test_module_and_local_dir_have_same_name(self, worker_id):
+        """If a pip-installed module and a local directory share a name, the module is collected."""
```

Comment:
It would be useful to keep a note here of why we want this (vs using the local dir) as I bet we'll ask ourselves in the future.

Reply:
Done in #2811!